File: packages/ogdf.js/src/layouts/module/HierarchyLayout.js
import createModule from "./module-helper";
import { PARAMETER_TYPE } from "../../utils/parameter-type";
const NAME = "HierarchyLayoutModule"

const MODULE_DIRECTORY = {
    FastHierarchyLayout: {
        fixedLayerDistance: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        layerDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 3.0
        },
        nodeDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 3.0
        }
    },
    FastSimpleHierarchyLayout: {
        layerDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 75
        },
        nodeDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 150
        },
        balanced: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        downward: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        leftToRight: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        }
    },
    OptimalHierarchyLayout: {
        fixedLayerDistance: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        layerDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 3.0
        },
        nodeDistance: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 3.0
        },
        weightBalancing: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 0.1
        },
        weightSegments: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 2.0
        }
    }
}

const HierarchyLayoutModule = createModule(NAME, MODULE_DIRECTORY)
export default HierarchyLayoutModule

File: packages/ogdf.js/src/module/src/HierarchyLayout.cpp
#include "../main.h"

#include <ogdf/layered/FastHierarchyLayout.h>
#include <ogdf/layered/FastSimpleHierarchyLayout.h>
#include <ogdf/layered/OptimalHierarchyLayout.h>

EM_PORT_API(FastHierarchyLayout *)
HierarchyLayoutModule_FastHierarchyLayout(bool fixedLayerDistance, double layerDistance, double nodeDistance)
{
    FastHierarchyLayout *fastHierarchy = new FastHierarchyLayout();
    fastHierarchy->fixedLayerDistance(fixedLayerDistance);
    fastHierarchy->layerDistance(layerDistance);
    fastHierarchy->nodeDistance(nodeDistance);
    return fastHierarchy;
}

EM_PORT_API(FastSimpleHierarchyLayout *)
HierarchyLayoutModule_FastSimpleHierarchyLayout(bool balanced, bool downward, double layerDistance, bool leftToRight, double nodeDistance)
{
    FastSimpleHierarchyLayout *fastSimpleHierarchy = new FastSimpleHierarchyLayout();
    fastSimpleHierarchy->balanced(balanced);
    fastSimpleHierarchy->downward(downward);
    fastSimpleHierarchy->layerDistance(layerDistance);
    fastSimpleHierarchy->leftToRight(leftToRight);
    fastSimpleHierarchy->nodeDistance(nodeDistance);
    return fastSimpleHierarchy;
}

EM_PORT_API(OptimalHierarchyLayout *)
HierarchyLayoutModule_OptimalHierarchyLayout(bool fixedLayerDistance, double layerDistance, double nodeDistance, double weightBalancing, double weightSegments)
{
    OptimalHierarchyLayout *optimalHierarchyLayout = new OptimalHierarchyLayout();
    optimalHierarchyLayout->fixedLayerDistance(fixedLayerDistance);
    optimalHierarchyLayout->layerDistance(layerDistance);
    optimalHierarchyLayout->nodeDistance(nodeDistance);
    optimalHierarchyLayout->weightBalancing(weightBalancing);
    optimalHierarchyLayout->weightSegments(weightSegments);
    return optimalHierarchyLayout;
}
